packages/ember/tests/component_registration_test.js

```diff
@@ -1,162 +1,177 @@
     assert.expect(1);
 
-    this.runTask(() => {
-      this.createApplication();
+    this.addTemplate('application', `<div id='wrapper'>{{#my-component}}{{text}}{{/my-component}}</div>`);
+    this.addTemplate('foo-bar-baz', 'No way!');
 
-      this.addTemplate('application', `<div id='wrapper'>{{#my-component}}{{text}}{{/my-component}}</div>`);
-      this.addTemplate('foo-bar-baz', 'No way!');
-
-      this.applicationInstance.register('controller:application', Controller.extend({
-        text: 'outer'
-      }));
-      this.applicationInstance.register('component:my-component', Component.extend({
-        text: 'inner',
-        layoutName: 'foo-bar-baz',
-        layout: this.compile('{{text}}-{{yield}}')
-      }));
+    this.application.instanceInitializer({
+      name: 'application-controller-layout',
+      initialize(applicationInstance) {
+        applicationInstance.register('controller:application', Controller.extend({
+          text: 'outer'
+        }));
+      }
+    });
+    this.application.instanceInitializer({
+      name: 'my-component-component-layout',
+      initialize(applicationInstance) {
+        applicationInstance.register('component:my-component', Component.extend({
+          text: 'inner',
+          layoutName: 'foo-bar-baz',
+          layout: compile('{{text}}-{{yield}}')
+        }));
+      }
     });
 
-    let text = this.$('#wrapper').text().trim();
-    assert.equal(text, 'inner-outer', 'The component is composed correctly');
+    this.visit('/').then(() => {
+      let text = this.$('#wrapper').text().trim();
+      assert.equal(text, 'inner-outer', 'The component is composed correctly');
+    });
   }
 
-  /*
-   * When an exception is thrown during the initial rendering phase, the
-   * `visit` promise is not resolved or rejected. This means the `applicationInstance`
-   * is never torn down and tests running after this one will fail.
-   *
-   * It is ugly, but since this test intentionally causes an initial render
-   * error, it requires globals mode to access the `applicationInstance`
-   * for teardown after test completion.
-   *
-   * Application "globals mode" is trigged by `autoboot: true`. It doesn't
-   * have anything to do with the resolver.
-   *
-   * We should be able to fix this by having the application eagerly stash a
-   * copy of each application instance it creates. When the application is
-   * destroyed, it can also destroy the instances (this is how the globals
-   * mode avoid the problem).
-   *
-   * See: https://github.com/emberjs/ember.js/issues/15327
-   */
   ['@test Using name of component that does not exist']() {
-    expectAssertion(() => {
-      this.runTask(() => {
-        this.createApplication();
+    this.addTemplate('application', `<div id='wrapper'>{{#no-good}} {{/no-good}}</div>`);
 
-        this.addTemplate('application', `<div id='wrapper'>{{#no-good}} {{/no-good}}</div>`);
-      });
+    expectAssertion(() => {
+      this.visit('/');
```

Comment:
I don't know exactly how to handle this case properly. Currently `expectAssertion` doesnt' deal with async, the prior `runTask` wrapping was forcing it to be "pseudo-sync"

Reply:
Would something like this work?
```js
expectAssertion(() => {
  this.runTask(() => this.visit('/'));
}, /.* named "no-good" .*/);
```
The test passes either way, so I'm not sure what the proper refactor here is.

Reply:
The test passes either way because `this.visit` is sync at the moment 😭 .

@mixonic reminded me that `expectAssertion(/some regexp/)` can be used to handle the async cases like this, maybe we could swap to that instead?

Reply:
Looked into this and although `expectDeprecation` has an async-friendly form `expectAssertion` does not currently. I think a `TODO: Use the async form of expectAssertion here when it is available` would be sufficient.